packages/cli/test/unit/util/dev/parse-query-string.test.ts

```diff
@@ -0,0 +1,46 @@
+import {
+  parseQueryString,
+  formatQueryString,
+} from '../../../../src/util/dev/parse-query-string';
+
+describe('parseQueryString', () => {
+  it('should parse to Map and format back to original String', async () => {
+    const querystring = '?a&b=1&c=2&c=3&d=&space%20bar=4&html=%3Ch1%3E';
+    const parsed = parseQueryString(querystring);
+    const map = new Map<string, string[]>([
+      ['a', []],
```

Comment:
I thought this case would have `[undefined]` as the value here? What if we're parsing `?a&a&a&a&a=1`?

Reply:
I also wonder if Map is even really the correct data structure here. How do we retain ordering when formatting again? i.e. `parse('?a=1&b=2&a=3'`)? I think the current code would format _both_ `a` values before the `b` value.

Reply:
That's a good question. I have used `Map` many times, but never with the intent of preserving key order, which I know `Object` does. Silly question, why not just use `URLSearchParams`?

Reply:
> The Map object holds key-value pairs and remembers the original insertion order of the keys. 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

Reply:
Although I think you might be right about the duplicate keys that are undefined. Let me add a test for that 👀

Reply:
The point about Map insertion order doesn't really address the issue though. My example had `a,b,a` and right now the code will format back to `a,a,b`.

Reply:
FYI, `URLSearchParams` allows for duplicate keys:

```
> const u = new URLSearchParams('?a=1&b=2&a=3');
undefined
> u
URLSearchParams { 'a' => '1', 'b' => '2', 'a' => '3' }
> u.entries()
URLSearchParams Iterator { [ 'a', '1' ], [ 'b', '2' ], [ 'a', '3' ] }
```

It also supports with or without the `?` and empty values:

```
> new URLSearchParams('a=&b=&a')
URLSearchParams { 'a' => '', 'b' => '', 'a' => '' }
```

Reply:
@cb1kenobi Unfortunetly this suffers from the same bug:

```
> new URLSearchParams('a=&b=&a')
URLSearchParams { 'a' => '', 'b' => '', 'a' => '' }
```

Because we need the first `a` to be empty string and the second `a` to be undefined. Otherwise we can't tell which one had an equals sign (see PR description).

I switched the code back to a plain Object so that its compatible with the existing `url.parse()` and `url.format()` API. 

I think the order of `a,b,a` to `a,a,b` will need to be fixed in a future PR since that is out of scope for the `vite dev` fix and that is the issue customers are blocked on.